aep/general/0155/aep.md

```diff
@@ -0,0 +1,109 @@
+identifier for particular requests. This can be useful for several purposes,
+such as:
+
+- De-duplicating requests from parallel processes
+- Ensuring the safety of retries
+- Auditing
+
+The purpose of idempotency keys is to provide idempotency guarantees: allowing
+the same request to be issued more than once without subsequent calls having
+any effect. In the event of a network failure, the client can retry the
+request, and the server can detect duplication and ensure that the request is
+only processed once.
+
+## Guidance
+
+APIs **may** add a `string idempotency_key` parameter to request messages
+(including those of standard methods) in order to uniquely identify particular
+requests.
+
+```proto
+message CreateBookRequest {
+  // The parent resource where this book will be created.
+  // Format: publishers/{publisher}
+  string parent = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference) = {
+      child_type: "library.example.com/Book"
+    }];
+
+  // The book to create.
+  Book book = 2 [(google.api.field_behavior) = REQUIRED];
+
+  // A unique identifier for this request. Restricted to 36 ASCII characters.
+  // A random UUID is recommended.
+  // This request is only idempotent if a `idempotency_key` is provided.
+  string idempotency_key = 3 [(google.api.field_info).format = UUID4];
+}
+```
+
+- Providing an idempotency key **must** guarantee idempotency.
+  - If a duplicate request is detected, the server **should** return the
+    response for the previously successful request, because the client most
+    likely did not receive the previous response.
+  - APIs **should** honor idempotency keys for at least an hour.
+- The `idempotency_key` field **must** be provided on the request message to
+  which it applies (and it **must not** be a field on resources themselves).
+- Idempotency keys **should** be optional.
+- Idempotency keys **should** be able to be UUIDs, and **may** allow UUIDs to
+  be the only valid format. The format restrictions for idempotency keys
+  **must** be documented.
+  - Idempotency keys that are UUIDs **must** be annotated with the
+    `google.api.FieldInfo.Format` value `UUID4` using the extension
+    `(google.api.field_info).format = UUID4`. See [AEP-202](../0202/aep.md) for
+    more.
+
+### Stale success responses
+
+In some unusual situations, it may not be possible to return an identical
+success response. For example, a duplicate request to create a resource may
+arrive after the resource has not only been created, but subsequently updated;
+because the service has no other need to retain the historical data, it is no
+longer feasible to return an identical success response.
+
+In this situation, the method **may** return the current state of the resource
+instead. In other words, it is permissible to substitute the historical success
+response with a similar response that reflects more current data.
```

Comment:
This behavior seems problematic, because such a response does not actually correlate with the request! What if the resource has subsequently been deleted, or made inaccessible? And even if the common case, how is a client to detect that they have received such a response?

Rather than silently replacing the no longer available appropriate response, I think it would be best to communicate that information. In HTTP terms, it would presumably be a [303 See Other](https://www.rfc-editor.org/rfc/rfc9110.html#section-15.4.4) response with a Location field identifying the URL to use for getting current state. And having provided that signal, it would then be possible _as an optimization_ to also include that representation along with a [Content-Location](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.7) field, e.g.
```
HTTP/1.1 303 See Other
…
Location: /path/to/resource
Content-Location: /path/to/resource

<current representation>
```

Reply:
Hmm, I don't know if 303 fits: "The 303 (See Other) status code indicates that the server is redirecting the user agent to a different resource..."

In the case you're talking about, it wouldn't be a different resource, but instead a different version of the same resource.  This isn't just semantics - a 303 is supposed to include a URI different than the current one.

But given the resource path, it's always possible to get the resource - we don't need a 303 (or anything else) to provide that info.

The main challenge I see is this case (which you mentioned):

1.  Update
2.  Update request succeeds but response lost
3.  Resource deleted
4.  Update again with same idempotency key

In that case, what's the right response to the second update request?  Arguably it's fine for it to 404; if the resource was deleted then the fact that the update succeeded probably doesn't matter (since update shouldn't have side effects, per https://google.aip.dev/134).

Reply:
For reference, the OASIS Repeatable Requests standard says
> The response returned by the server MUST be the success or failure state of the operation as first executed by the server ...

So returning the current state is only okay if the state hasn't changed since the operation was first executed. I think this design makes sense.

Reply:
I'm going to remove this section entirely from this PR - it's not clear it's needed given the new wording above about "equivalent responses".  But I'll create a separate PR adding this section back so that we can continue the discussion there.

Reply:
https://github.com/aep-dev/aep.dev/pull/78